Publishing diagnostics at startup fills up the channel

**Describe the bug**

I noticed an exception in the logs when testing the latest `clojure-lsp` binary. Not sure how this affects the editor, perhaps some of the diagnostics get discarded. I don't think I've seen this before.

**To Reproduce**
Steps to reproduce the behavior:
1. Open a large project (thousands of namespaces) in an editor
2. Look at the logs
3. Notice hundreds of "Publishing N diagnostics for file..." messages
4. See stacktrace

**Expected behavior**

No errors should occur. All diagnostics should be reported.

<details>
 <summary><b>Log - clojure-lsp</b></summary>
<pre>
2022-07-26T04:21:37.250Z  ERROR [taoensso.timbre:828] - Uncaught exception on thread: async-dispatch-2
com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine    PosixJavaThreads.java:  194
            com.oracle.svm.core.thread.JavaThreads.threadStartRoutine         JavaThreads.java:  597
                                                 java.lang.Thread.run              Thread.java:  829
                                                                  ...                               
   clojure.core.async.impl.concurrent/counted-thread-factory/reify/fn           concurrent.clj:   29
                   java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  628
                    java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1128
                                                                  ...                               
                                   clojure-lsp.handlers/initialize/fn             handlers.clj:  105
         clojure.core.async.impl.ioc-macros/run-state-machine-wrapped           ioc_macros.clj:  982
                 clojure.core.async.impl.ioc-macros/run-state-machine           ioc_macros.clj:  978
             clojure-lsp.handlers/initialize/fn/state-machine--auto--             handlers.clj:  105
          clojure-lsp.handlers/initialize/fn/state-machine--auto--/fn             handlers.clj:  106
             clojure-lsp.feature.diagnostics/publish-all-diagnostics!          diagnostics.clj:  165
            clojure-lsp.feature.diagnostics/sync-publish-diagnostics!          diagnostics.clj:  145
                                              clojure.core.async/put!                async.clj:  199
              clojure.core.async.impl.channels.ManyToManyChannel/put!             channels.clj:  156
java.lang.AssertionError: Assert failed: No more than 1024 pending puts are allowed on a single channel. Consider using a windowed buffer.
                          (< (.size puts) impl/MAX-QUEUE-SIZE)
</pre>
</details>

**User details (please complete the following information):**
 - OS: Arch Linux
 - Editor emacs
 - Version: `clojure-lsp 2022.07.24-18.25.43` + `clj-kondo 2022.06.23-SNAPSHOT`

I tried the latest nightly and the issue still remains

<details>
<pre>2022-08-12T10:03:27.652Z  INFO [clojure-lsp.dep-graph:300] - :maintain-dep-graph 0ms
2022-08-12T10:03:27.652Z  INFO [clojure-lsp.feature.java-interop:256] - [Java] JDK source cached loaded successfully.
2022-08-12T10:03:27.666Z  WARN [clojure-lsp.kondo:298] - Non-fatal error from clj-kondo: No configs copied.

2022-08-12T10:03:27.667Z  INFO [clojure-lsp.dep-graph:300] - :maintain-dep-graph 0ms
2022-08-12T10:03:27.667Z  INFO [clojure-lsp.handlers:116] - :did-open 45ms
2022-08-12T10:03:27.668Z  INFO [clojure-lsp.server:147] - :refreshing-test-tree 0ms
2022-08-12T10:03:27.795Z  INFO [clojure-lsp.handlers:?] - :code-lens 0ms
2022-08-12T10:03:27.835Z  ERROR [taoensso.timbre:828] - Uncaught exception on thread: async-thread-macro-8
com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine  PosixPlatformThreads.java:  202
            com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine       PlatformThreads.java:  705
                                                     java.lang.Thread.run                Thread.java:  829
                       java.util.concurrent.ThreadPoolExecutor$Worker.run    ThreadPoolExecutor.java:  628
                        java.util.concurrent.ThreadPoolExecutor.runWorker    ThreadPoolExecutor.java: 1128
                                                                      ...                                 
                                        clojure.core.async/thread-call/fn                  async.clj:  484
                                       clojure-lsp.handlers/initialize/fn               handlers.clj:   99
                 clojure-lsp.feature.diagnostics/publish-all-diagnostics!            diagnostics.clj:  165
                clojure-lsp.feature.diagnostics/sync-publish-diagnostics!            diagnostics.clj:  145
                                                  clojure.core.async/put!                  async.clj:  199
                  clojure.core.async.impl.channels.ManyToManyChannel/put!               channels.clj:  156
java.lang.AssertionError: Assert failed: No more than 1024 pending puts are allowed on a single channel. Consider using a windowed buffer.
                          (< (.size puts) impl/MAX-QUEUE-SIZE)
</pre>
</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -21,6 +21,7 @@
   - Avoid exceptions when clients use URIs that don't exist on disk.
   - Fix documentation on completion items. #1181
   - Fix rename of defrecords. #1165
+  - Fix to send all diagnostics to client at startup, even in very large projects. #1153
 
 ## 2022.07.24-18.25.43
 

--- a/common-test/src/clojure_lsp/test_helper.clj
+++ b/common-test/src/clojure_lsp/test_helper.clj
@@ -85,17 +85,17 @@
        (f)))))
 
 (defmacro let-mock-chans [bindings & body]
-  (assert (even? (count bindings)))
+  {:pre [(even? (count bindings))]}
   (let [bs (partition 2 bindings)
         let-bindings (mapcat (fn [[binding _]]
-                               `[~binding (async/chan 1)])
+                               [binding `(async/chan 1)])
                              bs)
-        alter-vars (map (fn [[binding chan-var]]
-                          `(alter-var-root ~chan-var (constantly ~binding)))
-                        bs)]
+        redef-bindings (mapcat (fn [[binding chan-var]]
+                                 [chan-var binding])
+                               bs)]
     `(let [~@let-bindings]
-       ~@alter-vars
-       ~@body)))
+       (with-redefs [~@redef-bindings]
+         ~@body))))
 
 (defn take-or-timeout [c timeout-ms]
   (let [timeout (async/timeout timeout-ms)

--- a/lib/src/clojure_lsp/feature/diagnostics.clj
+++ b/lib/src/clojure_lsp/feature/diagnostics.clj
@@ -141,38 +141,31 @@
         (not= :off depend-level)
         (concat (clj-depend-violations->diagnostics filename depend-level db))))))
 
-(defn sync-publish-diagnostics! [uri db]
-  (async/put! db/diagnostics-chan
-              {:uri uri
-               :diagnostics (find-diagnostics uri db)}))
+(defn publish-diagnostic!* [diagnostic]
+  (async/put! db/diagnostics-chan diagnostic))
 
-(defn async-publish-diagnostics! [uri db]
-  (if (#{:unit-test :api-test} (:env db)) ;; Avoid async on test which cause flakeness
-    (async/put! db/diagnostics-chan
-                {:uri uri
-                 :diagnostics (find-diagnostics uri db)})
-    (async/go
-      (async/>! db/diagnostics-chan
-                {:uri uri
-                 :diagnostics (find-diagnostics uri db)}))))
+(defn publish-all-diagnostics!* [diagnostics]
+  (async/onto-chan! db/diagnostics-chan diagnostics false))
+
+(defn publish-diagnostics! [uri db]
+  (publish-diagnostic!* {:uri uri
+                         :diagnostics (find-diagnostics uri db)}))
 
 (defn publish-all-diagnostics! [paths db]
-  (doseq [path paths]
-    (doseq [file (file-seq (io/file path))]
-      (let [filename (.getAbsolutePath ^java.io.File file)
-            uri (shared/filename->uri filename db)]
-        (when (not= :unknown (shared/uri->file-type uri))
-          (sync-publish-diagnostics! uri db))))))
+  (publish-all-diagnostics!*
+    (eduction (map io/file)
+              (mapcat file-seq)
+              (map #(.getAbsolutePath ^java.io.File %))
+              (map #(shared/filename->uri % db))
+              (remove #(= :unknown (shared/uri->file-type %)))
+              (map (fn [uri]
+                     {:uri uri
+                      :diagnostics (find-diagnostics uri db)}))
+              paths)))
 
-(defn publish-empty-diagnostics! [uri db]
-  (if (#{:unit-test :api-test} (:env db))
-    (async/put! db/diagnostics-chan
-                {:uri uri
-                 :diagnostics []})
-    (async/go
-      (async/>! db/diagnostics-chan
-                {:uri uri
-                 :diagnostics []}))))
+(defn publish-empty-diagnostics! [uri]
+  (publish-diagnostic!* {:uri uri
+                         :diagnostics []}))
 
 (defn ^:private unused-public-vars [var-defs kw-defs project-db kondo-config]
   (let [var-definitions (remove (partial exclude-public-diagnostic-definition? kondo-config) var-defs)

--- a/lib/src/clojure_lsp/feature/diagnostics.clj
+++ b/lib/src/clojure_lsp/feature/diagnostics.clj
@@ -141,10 +141,10 @@
         (not= :off depend-level)
         (concat (clj-depend-violations->diagnostics filename depend-level db))))))
 
-(defn publish-diagnostic!* [diagnostic]
+(defn ^:private publish-diagnostic!* [diagnostic]
   (async/put! db/diagnostics-chan diagnostic))
 
-(defn publish-all-diagnostics!* [diagnostics]
+(defn ^:private publish-all-diagnostics!* [diagnostics]
   (async/onto-chan! db/diagnostics-chan diagnostics false))
 
 (defn publish-diagnostics! [uri db]

--- a/lib/src/clojure_lsp/feature/file_management.clj
+++ b/lib/src/clojure_lsp/feature/file_management.clj
@@ -43,7 +43,7 @@
                  (-> state-db
                      (lsp.kondo/db-with-results kondo-result)
                      (lsp.depend/db-with-results depend-result))))
-    (f.diagnostic/async-publish-diagnostics! uri @db*))
+    (f.diagnostic/publish-diagnostics! uri @db*))
   (when allow-create-ns
     (when-let [create-ns-edits (create-ns-changes uri text @db*)]
       (async/>!! db/edits-chan create-ns-edits))))
@@ -162,9 +162,7 @@
             ;; https://github.com/clojure-lsp/clojure-lsp/issues/1027 and
             ;; https://github.com/clojure-lsp/clojure-lsp/issues/1028.
             (analyze-reference-filenames! filenames db*))
-          (let [db @db*]
-            (doseq [filename filenames]
-              (f.diagnostic/sync-publish-diagnostics! (shared/filename->uri filename db) db)))
+          (f.diagnostic/publish-all-diagnostics! filenames @db*)
           (producer/refresh-code-lens producer))))))
 
 (defn ^:private offsets [lines line character end-line end-character]
@@ -229,7 +227,7 @@
                                     (lsp.depend/db-with-results depend-result)
                                     (update :processing-changes disj uri)))
             (let [db @db*]
-              (f.diagnostic/sync-publish-diagnostics! uri db)
+              (f.diagnostic/publish-diagnostics! uri db)
               (when (settings/get db [:notify-references-on-file-change] true)
                 (notify-references filename old-db db components))
               (producer/refresh-test-tree producer [uri]))
@@ -264,7 +262,7 @@
 
 (defn ^:private file-deleted [db* uri filename]
   (swap! db* db-without-file uri filename)
-  (f.diagnostic/publish-empty-diagnostics! uri @db*))
+  (f.diagnostic/publish-empty-diagnostics! uri))
 
 (defn did-change-watched-files [changes db*]
   (doseq [{:keys [uri type]} changes]

--- a/lib/src/clojure_lsp/handlers.clj
+++ b/lib/src/clojure_lsp/handlers.clj
@@ -95,8 +95,7 @@
           components)
         (let [db @db*]
           (when (settings/get db [:lint-project-files-after-startup?] true)
-            (async/thread
-              (f.diagnostic/publish-all-diagnostics! (-> db :settings :source-paths) db)))
+            (f.diagnostic/publish-all-diagnostics! (-> db :settings :source-paths) db))
           (async/go
             (f.clojuredocs/refresh-cache! db*))
           (async/go

--- a/lib/test/clojure_lsp/features/diagnostics_test.clj
+++ b/lib/test/clojure_lsp/features/diagnostics_test.clj
@@ -291,7 +291,7 @@
                   (foo 1 ['a 'b])
                   (foo 1 2 3 {:k 1 :v 2})"]
         (h/let-mock-chans
-          [mock-diagnostics-chan #'db/diagnostics-chan]
+          [mock-diagnostics-chan db/diagnostics-chan]
           (h/load-code-and-locs code)
           (let [{:keys [uri diagnostics]} (h/take-or-timeout mock-diagnostics-chan 500)]
             (is (= "file:///a.clj" uri))
@@ -326,7 +326,7 @@
                     (foo 1)
                     (bar))"]
         (h/let-mock-chans
-          [mock-diagnostics-chan #'db/diagnostics-chan]
+          [mock-diagnostics-chan db/diagnostics-chan]
           (h/load-code-and-locs code)
           (let [{:keys [uri diagnostics]} (h/take-or-timeout mock-diagnostics-chan 500)]
             (is (= "file:///a.clj" uri))
@@ -346,7 +346,7 @@
                   (bar :a)
                   (bar :a :b)"]
         (h/let-mock-chans
-          [mock-diagnostics-chan #'db/diagnostics-chan]
+          [mock-diagnostics-chan db/diagnostics-chan]
           (h/load-code-and-locs code)
           (let [{:keys [uri diagnostics]} (h/take-or-timeout mock-diagnostics-chan 500)]
             (is (= "file:///a.clj" uri))
@@ -362,7 +362,7 @@
                   (foo 1 2)
                   (foo 1)"]
         (h/let-mock-chans
-          [mock-diagnostics-chan #'db/diagnostics-chan]
+          [mock-diagnostics-chan db/diagnostics-chan]
           (h/load-code-and-locs code)
           (let [{:keys [uri diagnostics]} (h/take-or-timeout mock-diagnostics-chan 500)]
             (is (= "file:///a.clj" uri))
@@ -372,7 +372,7 @@
   (testing "custom unused namespace declaration"
     (h/clean-db!)
     (h/let-mock-chans
-      [mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-diagnostics-chan db/diagnostics-chan]
       (h/load-code-and-locs "(ns foo.bar)" (h/file-uri "file:///foo/bar.clj"))
       (let [{:keys [uri diagnostics]} (h/take-or-timeout mock-diagnostics-chan 500)]
         (is (= "file:///foo/bar.clj" uri))

--- a/lib/test/clojure_lsp/features/file_management_test.clj
+++ b/lib/test/clojure_lsp/features/file_management_test.clj
@@ -27,7 +27,7 @@
   (h/load-code-and-locs "(ns some-jar)" (h/file-uri "file:///some/path/to/jar.jar:/some/file.clj"))
   (testing "when file exists on disk"
     (h/let-mock-chans
-      [mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-diagnostics-chan db/diagnostics-chan]
       (with-redefs [shared/file-exists? (constantly true)]
         (f.file-management/did-close "file:///user/project/src/clj/foo.clj" db/db*))
       (is (get-in @db/db* [:analysis "/user/project/src/clj/foo.clj"]))
@@ -38,7 +38,7 @@
       (h/assert-no-take mock-diagnostics-chan 500)))
   (testing "when local file not exists on disk"
     (h/let-mock-chans
-      [mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-diagnostics-chan db/diagnostics-chan]
       (with-redefs [shared/file-exists? (constantly false)]
         (f.file-management/did-close "file:///user/project/src/clj/bar.clj" db/db*))
       (is (nil? (get-in @db/db* [:analysis "/user/project/src/clj/bar.clj"])))
@@ -51,7 +51,7 @@
              (h/take-or-timeout mock-diagnostics-chan 500)))))
   (testing "when file is external we do not remove analysis"
     (h/let-mock-chans
-      [mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-diagnostics-chan db/diagnostics-chan]
       (with-redefs [shared/file-exists? (constantly false)]
         (f.file-management/did-close "file:///some/path/to/jar.jar:/some/file.clj" db/db*))
       (is (get-in @db/db* [:analysis "/some/path/to/jar.jar:/some/file.clj"]))
@@ -64,8 +64,8 @@
 (deftest did-open
   (testing "on an empty file"
     (h/let-mock-chans
-      [mock-edits-chan #'db/edits-chan
-       mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-edits-chan db/edits-chan
+       mock-diagnostics-chan db/diagnostics-chan]
       (let [filename "/user/project/src/aaa/bbb.clj"
             uri (h/file-uri (str "file://" filename))]
         (swap! db/db* shared/deep-merge {:settings {:auto-add-ns-to-new-files? true
@@ -92,7 +92,7 @@
 
 (deftest did-change
   (h/let-mock-chans
-    [mock-changes-chan #'db/current-changes-chan]
+    [mock-changes-chan db/current-changes-chan]
     (let [original-text (h/code "(ns aaa)"
                                 "(def foo 1)")
           edited-text (h/code "(ns aaa)"
@@ -112,7 +112,7 @@
 (deftest did-change-watched-files
   (testing "created file"
     (h/let-mock-chans
-      [mock-created-chan #'db/created-watched-files-chan]
+      [mock-created-chan db/created-watched-files-chan]
       (f.file-management/did-change-watched-files
         [{:type :created
           :uri h/default-uri}]
@@ -120,7 +120,7 @@
       (is (= h/default-uri (h/take-or-timeout mock-created-chan 500)))))
   (testing "deleted file"
     (h/let-mock-chans
-      [mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-diagnostics-chan db/diagnostics-chan]
       (f.file-management/did-change-watched-files
         [{:type :deleted
           :uri h/default-uri}]

--- a/lib/test/clojure_lsp/features/resolve_macro_test.clj
+++ b/lib/test/clojure_lsp/features/resolve_macro_test.clj
@@ -3,7 +3,6 @@
    [clojure-lsp.db :as db]
    [clojure-lsp.feature.resolve-macro :as f.resolve-macro]
    [clojure-lsp.test-helper :as h]
-   [clojure.core.async :as async]
    [clojure.test :refer [deftest is testing]]))
 
 (h/reset-db-after-test)
@@ -26,7 +25,6 @@
       (is (not (f.resolve-macro/find-full-macro-symbol-to-resolve (h/zloc-from-code b-code) (h/file-uri "file:///a.clj") @db/db*))))))
 
 (deftest resolve-macro-as
-  (alter-var-root #'db/diagnostics-chan (constantly (async/chan 1)))
   (let [code (h/code "(ns some-ns)"
                      "(defmacro foo [name] `(def ~name))"
                      "(foo my|-fn)"

--- a/lib/test/clojure_lsp/handlers_test.clj
+++ b/lib/test/clojure_lsp/handlers_test.clj
@@ -28,7 +28,7 @@
   (testing "opening a existing file"
     (h/clean-db!)
     (h/let-mock-chans
-      [mock-diagnostics-chan #'db/diagnostics-chan]
+      [mock-diagnostics-chan db/diagnostics-chan]
       (let [_ (h/load-code-and-locs "(ns a) (when)")]
         (is (some? (get-in @db/db* [:analysis (h/file-path "/a.clj")])))
         (let [{:keys [uri diagnostics]} (h/take-or-timeout mock-diagnostics-chan 500)]
@@ -44,7 +44,7 @@
                          :client-capabilities {:workspace {:workspace-edit {:document-changes true}}}
                          :project-root-uri (h/file-uri "file:///project")})
     (h/let-mock-chans
-      [mock-edits-chan #'db/edits-chan]
+      [mock-edits-chan db/edits-chan]
       (h/load-code-and-locs "" (h/file-uri "file:///project/src/foo/bar.clj"))
       (h/assert-submaps
         [{:edits [{:range {:start {:line 0, :character 0}
@@ -59,7 +59,7 @@
                          :client-capabilities {:workspace {:workspace-edit {:document-changes true}}}
                          :project-root-uri (h/file-uri "file:///project")})
     (h/let-mock-chans
-      [mock-edits-chan #'db/edits-chan]
+      [mock-edits-chan db/edits-chan]
       (h/load-code-and-locs "" (h/file-uri "file:///project/src/foo/baz.edn"))
       (h/assert-no-take mock-edits-chan 500)
       (is (some? (get-in @db/db* [:analysis (h/file-path "/project/src/foo/baz.edn")]))))))

